src/network.py

```diff
@@ -94,44 +94,53 @@ def __init__(self, text_m, speech_m, discriminator=None, teacher=None):
         self.discriminator = discriminator
         self.teacher = teacher
 
-    def text_ae(self, text, text_len):
-        return self.text_m.forward(text, text_len, noise_in=True, teacher_ratio=self.teacher.get_val())
-    
-    def speech_ae(self, mel, mel_len):
-        return self.speech_m.forward(mel, mel_len, noise_in=True, teacher_ratio=self.teacher.get_val())
+    def text_ae(self, text, text_len, ret_enc_hid=False):
+        return self.text_m.forward(text, text_len, noise_in=True, teacher_ratio=self.teacher.get_val(), ret_enc_hid=ret_enc_hid)
 
-    def cm_text_in(self, text, text_len):
+    def speech_ae(self, mel, mel_len, ret_enc_hid=False):
+        return self.speech_m.forward(mel, mel_len, noise_in=True, ret_enc_hid=ret_enc_hid, teacher_ratio=self.teacher.get_val())
+
+    def cm_text_in(self, text, text_len, ret_enc_hid=False):
         t_e_o, t_mask = self.text_m.encode(text, text_len)
-        pred, _, pred_lens = self.speech_m.infer_sequence(t_e_o, t_mask)
-        cm_s_e_o, cm_mask = self.speech_m.encode(pred, pred_lens)
+        _, post_pred, _, pred_lens = self.speech_m.infer_sequence(t_e_o, t_mask)
+        cm_s_e_o, cm_mask = self.speech_m.encode(post_pred, pred_lens)
         text_pred = self.text_m.decode_sequence(text, text_len, cm_s_e_o, cm_mask,
             teacher_ratio=self.teacher.get_val())
+        if ret_enc_hid:
+            return text_pred, t_e_o[0], cm_s_e_o[0]
         return text_pred
 
-    def cm_speech_in(self, mel, mel_len):
+    def cm_speech_in(self, mel, mel_len, ret_enc_hid=False):
         s_e_o, s_mask = self.speech_m.encode(mel, mel_len)
         text_pred, text_pred_len = self.text_m.infer_sequence(s_e_o, s_mask)
         cm_t_e_o, cm_t_masks = self.text_m.encode(text_pred, text_pred_len)
-        pred, stop_pred = self.speech_m.decode_sequence(mel, mel_len, cm_t_e_o, 
+        pre_pred, post_pred, stop_pred = self.speech_m.decode_sequence(mel, mel_len, cm_t_e_o,
             cm_t_masks, teacher_ratio=self.teacher.get_val())
-        return pred, stop_pred
-    
-    def tts(self, text, text_len, mel, mel_len, infer=False):
+        if ret_enc_hid:
+            return pre_pred, post_pred, stop_pred, s_e_o[0], cm_t_e_o[0]
```

Comment:
I don't think s_e_o[0] will correctly fetch the hidden state.  encoder outputs should be [batch_size x seq_len x hidden_dim]!

Reply:
encoder outputs is a tuple   


 def encode(self, text, text_lens, noise_in=False):
        """
        Returns:
            - output: a tensor of shape [batch_size x seq_len x latent_dim], which
                is the projected hidden state from each timestep in the sequence
            - h_t: the (h_n, c_n) pair from the last timestep of the network.
                Dimension varies depending on num_layers, but should
                be proportional to  [num_layers, batch, hidden_size]
            - pad_mask: a tensor of shape [batch_size x seq_len] with 1s for padded
                indices and 0 for everything else
        """
        embedded_phonemes, pad_mask = self.preprocess(text, text_lens, noise_in)
        enc_output, hidden_state = self.encoder(embedded_phonemes)
        return (hidden_state, enc_output), pad_mask

Reply:
Ah, I was still thinking of transformers where this will be  [batch_size x seq_len x hidden_dim].   note for when we add discriminator support for trasformers then!